### ISSUE TITLE
[Deepspeed] [performance] inefficient load with `from_pretrained` w/ zero3

### ISSUE THREAD

**Issue by stas00** [huggingface/transformers#12273]
# 🚀 Feature request

Currently under Deepspeed stage3 with `from_pretrained` we:

a. loop over each sub-module in zero.Init
1. init the sub-module
2. shard and scatter the shards

b. then to load pre-trained weights we loop over each sub-module:
1. gather the shards
2. `load_state_dict` for the one layer layer
3. shard and scatter the shards

c. any sub-module params that weren't in the pretrained state_dict
1. run the postponed `module_init` as it was done in https://github.com/huggingface/transformers/pull/11471
2. shard and scatter the shards XXX: I actually don't think zero.Init was handled here. need to fix that https://github.com/huggingface/transformers/issues/12272

which takes much longer than just 

a. init the modules w/o allocating any storage as it has been implemented in pt-1.9.0/1.9.1 https://pytorch.org/tutorials/prototype/skip_param_init.html#implementation-details

b. for each sub-module:
1. load_state_dict 
2. shard and scatter the shards

c. any sub-module params that weren't in the pretrained state_dict
1. materialize and module_init
2. shard and scatter the shards

Solving this will most likely require support from Deepspeed, or perhaps we can just try to remove `zero.Init` if the weights aren't materialized during model creation. So the very first sharding will get postponed to the `load_state_dict` stage.

**Comment by qgallouedec** [comment]
Hey @stas00, apologies for the (very) late reply, we're catching up on the backlog.

This WIP tracking issue predates a full rewrite of the DeepSpeed ZeRO-3 loading path, so the specific flow it describes no longer exists. `zero.Init` now runs under a `no_init_weights()` context, see https://github.com/huggingface/transformers/blob/150eb7c9ed4091294c829fa0e9466b090cb0f87f/src/transformers/modeling_utils.py#L1646-L1673
so parameters are partitioned as empty placeholders without materializing/initializing weights on the first scatter.

On load, `_load_state_dict_into_zero3_model` gathers only the params actually present in each checkpoint shard, then loads and re-partitions them https://github.com/huggingface/transformers/blob/150eb7c9ed4091294c829fa0e9466b090cb0f87f/src/transformers/integrations/deepspeed.py#L503-L509
The blanket scatter/gather/scatter is gone, and the postponed `module_init` from #11471 was replaced by `_is_hf_initialized` tracking, which also covers the gap noted in #12272.

Closing as outdated

**Comment by stas00** [comment]
yes, this is a really old ticket. I appreciate the commentary, @qgallouedec
